homework/hw6/task2_oop2.py

```diff
@@ -0,0 +1,190 @@
+    homework - для объекта Homework, если передан не этот класс -  выкинуть
+    подходящие по смыслу исключение с сообщением:
+    'You gave a not Homework object'
+    solution - хранит решение ДЗ как строку
+    author - хранит объект Student
+    created - c точной датой и временем создания
+2. Если задание уже просрочено хотелось бы видеть исключение при do_homework,
+а не просто принт 'You are late'.
+Поднимайте исключение DeadlineError с сообщением 'You are late' вместо print.
+3. Student и Teacher имеют одинаковые по смыслу атрибуты
+(last_name, first_name) - избавиться от дублирования с помощью наследования
+4.
+Teacher
+Атрибут:
+    homework_done - структура с интерфейсом как в словаря, сюда поподают все
+    HomeworkResult после успешного прохождения check_homework
+    (нужно гаранитровать остутствие повторяющихся результатов по каждому
+    заданию), группировать по экземплярам Homework.
+    Общий для всех учителей. Вариант ипользования смотри в блоке if __main__...
+Методы:
+    check_homework - принимает экземпляр HomeworkResult и возвращает True если
+    ответ студента больше 5 символов, так же при успешной проверке добавить в
+    homework_done.
+    Если меньше 5 символов - никуда не добавлять и вернуть False.
+    reset_results - если передать экземпряр Homework - удаляет только
+    результаты этого задания из homework_done, если ничего не передавать,
+    то полностью обнулит homework_done.
+PEP8 соблюдать строго.
+Всем перечисленным выше атрибутам и методам классов сохранить названия.
+К названием остальных переменных, классов и тд. подходить ответственно -
+давать логичные подходящие имена.
+"""
+import datetime
+from collections import defaultdict
+from typing import NoReturn, Optional, Type, TypeVar, Union
+
+T = TypeVar("T")
+
+
+class NotHomework(Exception):
+    """Exception raised when you put non-Homework object."""
+
+    def __init__(self, message: str = "You gave not a homework object"):
+        self.message = message
+        super().__init__(self.message)
+
+
+class DeadlineError(Exception):
+    """Exception raised when you are late."""
+
+    def __init__(self, message: str = "You are late"):
+        self.message = message
+        super().__init__(self.message)
+
+
+class Person:
+    """Person."""
+
+    def __init__(self, last_name: str, first_name: str):
+        self.last_name = last_name
+        self.first_name = first_name
+
+
+class Homework:
+    """Homework."""
+
+    def __init__(self, text: str, deadline: int):
+        self.text = text
+        self.created = datetime.datetime.today()
+        self.deadline = datetime.timedelta(deadline)
+
+    def is_active(self) -> bool:
+        """Active.
+
+        Checks if this homework is still legitimate to do. True if so.
+        """
+        return datetime.datetime.today() <= self.deadline + self.created
+
+
+class Student(Person):
+    """Student."""
+
+    def do_homework(self, todo: Homework, solution: str) -> Union[T, DeadlineError]:
```

Comment:
Here we should use NoReturn, because function doesn't return anything in case of exception raising